### PR TITLE
Add data split mode to DMatrix MetaInfo

### PR DESCRIPTION
**Pull request by rongou**
We need this information to change the behavior of the predictor when data is split column-wise.

Part of #8424

### CODE REVIEW

**Comment by rongou** [comment]
@trivialfis

**Review by trivialfis — COMMENTED** [review]
Do we still need the one in learner if DMatrix contains all the information?

**Comment by rongou** [comment]
Hmm, if we don't declare a training parameter, how would the user specify it? I imagine if someone wants to do column-wise split, they'd do something like:
```python
param = { 'tree_method': 'gpu_hist', data_split_mode: 'col'}
xgb.train(param, ...)
```

**Comment by trivialfis** [comment]
If the DMatrix contains the information the user doesn't need to specify that parameter right?

**Comment by rongou** [comment]
@trivialfis this is what it looks like if we take out the `dsplit` parameter out of training and put it in data loading. Let me know what you think. Still need to fix some downstream calls.

**Comment by trivialfis** [comment]
I think the DMatrix one is better.

A little bit of background here, @rongou shared some potential changes with me offline for federated learning. One important factor is how we handle the data split parameter in the predictor. The original plan was to pass it as part of the learner train param or learner model param. We later thought it might be possible to make it part of the DMatrix, and all algorithms will dispatch based on the DMatrix it receives, which from my perspective is cleaner than the learner option.

cc @RAMitchell .

**Comment by rongou** [comment]
@trivialfis Turns out this actually solves a problem we talked about before. If the user has done some preprocessing to split the training data beforehand, now they can pass in `data_split_mode='none'` to avoid further splitting.

Other than that bit of additional functionality, this PR preserves the existing behavior. Do you think we should merge it? The CI should pass now.

**Review by trivialfis — COMMENTED** [review]
I agree on using DMatrix as the parameter holder. Some other questions in comments.

File: include/xgboost/c_api.h [hunk 1]
```diff
@@ -128,10 +128,11 @@ XGB_DLL int XGBGetGlobalConfig(char const **out_config);
  * \brief load a data matrix
  * \param fname the name of the file
  * \param silent whether print messages during loading
+ * \param dsplit data split mode
  * \param out a loaded data matrix
  * \return 0 when success, -1 when failure happens
  */
-XGB_DLL int XGDMatrixCreateFromFile(const char *fname, int silent, DMatrixHandle *out);
+XGB_DLL int XGDMatrixCreateFromFile(const char *fname, int silent, int dsplit, DMatrixHandle *out);
```
[the remark this answers is not in the document]
Comment (rongou) [reply include/xgboost/c_api.h:135]: This is the only place I'm aware of that's dealing with data split. Do those third party libraries support splitting?

If we take a concrete example, say, the Higgs dataset, here are the steps to get to training:
* Download the dataset from UCI.
* Unzip to csv.
* Manually split the csv to train/test (e.g. 90%/10%).

Current behavior:
* Local training: feed the train/test csv files to `DMatrix`, not additional splitting happens.
* Distributed training: feed the train/test csv files to `DMatrix`, xgboost automatically splits both files by the number of workers.
* Federated learning: we need to manually split the train/test files first to simulate the federated environment. For each worker, feed the already split train/test files to `DMatrix`, no additional splitting happens.

With this PR:
* Local training: no behavior change.
* Distributed training: feed the train/test csv files to `DMatrix`, xgboost automatically splits both files by the number of workers. There is an additional option to manually split the files first, then create the `DMatrix` with `data_split_mode='none'`.
* Federated learning: no behavior change.

Future:
* Local training: no change.
* Distributed training: support column-wise split by passing in `data_split_mode='col'` to the `DMatrix`.
* Federated learning: support vertical learning by passing in, say, `federated_data_partition=vertical` to the communicator. We'd assume the `DMatrix` is still created with `data_split_mode='none'`.

Does that make more sense?

I'll add a new API to take in JSON and support the data split mode.

[the remark this answers is not in the document]
Comment (trivialfis) [reply include/xgboost/c_api.h:135]: Thank you for the explanation. My concern of using the file parser is that I think we will eventually integrate xgboost with some sort of federated learning framework that manages the data just like dask/spark for distributed computing. In that case, the file parser route is less ideal.

[the remark this answers is not in the document]
Comment (rongou) [reply include/xgboost/c_api.h:135]: Thought about this more. I think we do have two separate concepts mixed together. Perhaps it's better to separate them out. Let's look at each training scenario.

**Local training**: single input only.

**Distributed training with non-file api**: assuming we are using a non-file api (spark/cudf etc.), and data is already split during preprocessing. We still need to tell xgboost how the data is split. We can add a field to `MetaInfo`, call it `data_distribution`? Which defaults to `row`, but can also be `column`, maybe later `block`.

**Federated learning**: here data is already siloed between workers, so we can rely on the same `MetaInfo.data_distribution` field.

Finally, the **`XGDMatrixCreateFromFile` API**: probably not very useful for production, but helpful for quick benchmarking and ease of use (just point `DMatrix` at a file). For local and federated training, it's single input only. For distributed training, currently by default it splits the input by row. We want to be able to pass in a parameter so it splits by column. We can reuse the existing `data_split_mode` enum.

In summary, here is the proposal:
* Add a `data_distribution` field to `MetaInfo` with values:
  - row (default)
  - column
  - block (for future use; can leave it out for now)
* Deprecate the current c api  `XGDMatrixCreateFromFile`.
* Add a new c api `XGDMatrixCreateFromFile` which takes in a JSON config that includes the `data_split_mode` field.
* Don't change jvm/R APIs for now, but add a new `data_distribution` parameter to the python `DMatrix` class.

Does that sound reasonable?

@hcho3 since he worked on block distribution.

[the remark this answers is not in the document]
Comment (trivialfis) [reply include/xgboost/c_api.h:135]: I think adding a parameter to DMatrix is possible but would be great if it's limited to the file API and only at C API level (instead of an attribute of DMatrix). For instance, it can be a parameter in the input URI:
``` python
DMatrix("data.csv?need_split=1", data_split_mode="column")
```

[the remark this answers is not in the document]
Comment (rongou) [reply include/xgboost/c_api.h:135]: But we still need an indicator somewhere to know how the data is split, since the training behavior is different for different split modes.

[the remark this answers is not in the document]
Comment (trivialfis) [reply include/xgboost/c_api.h:135]: We can let `data_split_mode` be the indicator for how the data is split, and use the `need_split` in the URI to tell DMatrix whether it should slice itself.

[the remark this answers is not in the document]
Comment (rongou) [reply include/xgboost/c_api.h:135]: But ultimately `data_split_mode` needs to be in `MetaInfo`, right? With your example
```
DMatrix("data.csv?need_split=1", data_split_mode="column")
```
don't we still need to pass in the `data_split_mode` through the c api?

[the remark this answers is not in the document]
Comment (trivialfis) [reply include/xgboost/c_api.h:135]: Yup, apologies. We still need the deprecation

[the remark this answers is not in the document]
Comment (trivialfis) [reply include/xgboost/c_api.h:135]: What do you think? ;-)

[the remark this answers is not in the document]
Comment (rongou) [reply include/xgboost/c_api.h:135]: Sounds good. We can probably just pass `need_split` as part of the JSON request.

[the remark this answers is not in the document]
Comment (rongou) [reply include/xgboost/c_api.h]: Added `XGDMatrixCreateFromFileV2` for lack of a better name. Ended up not exposing `need_split` and just replicating the current behavior.

[the remark this answers is not in the document]
Comment (rongou) [reply include/xgboost/c_api.h]: Ok added `need_split` to the JSON config, but for now it's not exposed in the python api. We can always add it later if needed.

File: include/xgboost/c_api.h [hunk 2]
```diff
@@ -126,12 +126,29 @@ XGB_DLL int XGBGetGlobalConfig(char const **out_config);
 
 /*!
  * \brief load a data matrix
+ * \deprecated since 1.7.3
```
Comment (trivialfis) [inline include/xgboost/c_api.h:129]: I think the next big one would be 2.0 unless something major comes up.

File: include/xgboost/c_api.h [hunk 3]
```diff
@@ -126,12 +126,29 @@ XGB_DLL int XGBGetGlobalConfig(char const **out_config);
...
+ *   - silent (optional): Whether to print message during loading. Default to true.
+ *   - need_split (optional): Whether to split the file. Default to true in distributed mode, false
+ *     otherwise.
+ *   - data_split_mode (optional): Whether to split by row or column. If need_split is true, the
+ *     file is split accordingly; if false, this is only an indicator on how the file was split
+ *     beforehand. Default to row.
+ * \param out a loaded data matrix
+ * \return 0 when success, -1 when failure happens
+ */
+XGB_DLL int XGDMatrixCreateFromFileV2(char const *config, DMatrixHandle *out);
```
Comment (trivialfis) [inline include/xgboost/c_api.h:150]: Have you considered making it a proper API for URI? For instance `FromURI`? Since users actually need to use some of the URI formats like `file.txt?format=csv`. XGBoost/dmlc-core doesn't guess the format and is a source of bugs when users pass in only the file name.

Also, do you plan to introduce the `need_split` with other input sources as well? If not, we can make it an URI parameter and limit its use to this function. Otherwise, we need to have an additional parameter in all language bindings.

**Review by trivialfis — COMMENTED** [review]
The implementation looks good to me. Excellent work! But questions regarding the interface and the scope of the new parameter.

File: include/xgboost/c_api.h (hunk 2 above)
Comment (rongou) [reply include/xgboost/c_api.h]: Done.

File: include/xgboost/c_api.h (hunk 3 above)
Comment (rongou) [reply include/xgboost/c_api.h]: Removed it as a parameter.

**Comment by rongou** [comment]
@trivialfis finally got the CI to pass. PTAL

**Review by trivialfis — APPROVED** [review]
Thank you for the work on preparing the DMatrix for column split!
